src/ai/backend/common/service_discovery/redis_discovery/service_discovery.py

```diff
@@ -0,0 +1,105 @@
+import json
+import uuid
+from dataclasses import dataclass
+from typing import Any, Sequence
+
+from redis.asyncio.client import Pipeline, Redis
+
+from ai.backend.common import redis_helper
+from ai.backend.common.json import dump_json_str
+from ai.backend.common.types import RedisConnectionInfo
+
+from ..service_discovery import ServiceDiscovery, ServiceMetadata
+
+_DEFAULT_PREFIX = "service_discovery"
+
+
+@dataclass
+class RedisServiceDiscoveryArgs:
+    redis: RedisConnectionInfo
+    ttl: int = 60
+    prefix: str = _DEFAULT_PREFIX
+
+
+class RedisServiceDiscovery(ServiceDiscovery):
+    _redis: RedisConnectionInfo
+    _ttl: int = 60
+    _prefix: str
+
+    def __init__(self, args: RedisServiceDiscoveryArgs) -> None:
+        self._redis = args.redis
+        self._ttl = args.ttl
+        self._prefix = args.prefix
+
+    async def _hsetex(self, key: str, mapping: dict[str, Any]) -> None:
+        # TODO: Use actual `hsetex` after upgrading to redis-py 8.0
+        async def _pipe_builder(r: Redis) -> Pipeline:
+            pipe = r.pipeline()
+            await pipe.hset(key, mapping={k: dump_json_str(v) for k, v in mapping.items()})
+            await pipe.expire(key, self._ttl)
+            await pipe.execute()
+            return pipe
+
+        await redis_helper.execute(self._redis, _pipe_builder)
```

Comment:
Now, we can set expiry in redis service discovery, so I think `_sweep_unhealthy_service` in the service discovery loop should only be used in etcd service discovery.